No more syntex

fitzgen opened the pull request:
There are a few commits in this PR, but the big one is the commit that goes `syntex` -> `quote` for code generation. I've included a copy of its commit message below.

I tried to verify that it works with the stylo build still, but there were some issues, and then I checked with master, and that wasn't working either. So now I'm C-Reducing the failures on master and figured that this is at least ready for feedback in the meantime.

r? @emilio 

----------------------------

The `syntex` crate is unmaintained. It is slow to build, and additionally it requires that we pre-process `src/codegen/mod.rs` before we build the `bindgen` crate.

The `quote` crate provides similar quasi-quoting functionality, is maintained, and builds faster. It doesn't have a typed API or builders, however; it only deals with tokens.

Before this commit:

```
$ cargo clean; cargo build
<snip>
    Finished dev [unoptimized + debuginfo] target(s) in 98.75 secs
```

After this commit:

```
$ cargo clean; cargo build
<snip>
    Finished dev [unoptimized + debuginfo] target(s) in 46.26 secs
```

Build time is cut in half! But what about run time?

Before this commit:

```
Generated Stylo bindings in: Duration { secs: 3, nanos: 521105668 }
```

After this commit:

```
Generated Stylo bindings in: Duration { secs: 3, nanos: 548797242 }
```

So it appears to be about 20ms slower at generating Stylo bindings, but I suspect this is well within the noise.

Finally, this also lets us remove that nasty `mem::transmute` inside `bindgen::ir::BindgenContext::gen` that was used for the old `syntex` context. Now `BindgenContext` doesn't have a lifetime parameter either. This should make it easier to revisit doing our analyses in parallel with `rayon`, since that context was one of the things that made it hard for `BindgenContext` to implement `Sync`.

Fixes #925

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

emilio commented on src/codegen/helpers.rs:
```diff
@@ -1,66 +1,62 @@
+    pub fn link_name(name: &str) -> quote::Tokens {
+        // This is ridiculous, but rustfmt doesn't seem to be formatting
+        // `link_name` attributes very well, so we jump through these formatting
+        // hoops to make it work.
```
Can you file an issue upstream and reference it here?

fitzgen replied:
I've filed a handful of bugs upstream, and they've all been fixed now, so this comment might actually be out of date. I'll double check.

emilio commented on src/codegen/helpers.rs:
```diff
@@ -116,64 +119,50 @@ pub mod ast_ty {
-                .lit()
-                .byte_str(string),
-        )
+        let b = quote::ByteStr(&string);
```
ooh, nice

fitzgen commented:
> If I understand it well, we're running rustfmt on the bindings now, and we need to have the latest rustfmt-nightly installed locally to run tests so that they pass on CI.
> 
> Is that right? If so, I think we should document that.

Yes, it will be automatically installed by the `Once` callback in the `rustfmt` function in the tests file.

I'll document it in `CONTRIBUTING.md` as well.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #926) made this pull request unmergeable. Please resolve the merge conflicts.

fitzgen commented:
* Added note about `rustfmt` in `CONTRIBUTING.md`
* Removed the funky formatting for `link_name` -- new `rustfmt` has fixed the bug we were running into

re-r? @emilio

fitzgen commented:
Oops, just saw I need to rebase now.

fitzgen commented:
Ok, rebased.

emilio commented on CONTRIBUTING.md:
````diff
@@ -96,6 +96,14 @@ There is also the `./bindgen-integration` crate, which uses `bindgen` to
+
+```
+$ rustup update nightly
+$ rustup run nightly cargo install -f rustfmt-nightly
````
What happens when `rustfmt` changes the default formatting? All our CI breaks until we update them? That seems not great :(

fitzgen replied:
Nope, because both the generated bindings *and* the expectations are run through `rustfmt` before being compared.

emilio commented on tests/tests.rs:
```diff
@@ -5,42 +5,132 @@ extern crate shlex;
+        // Because `rustfmt` needs to match its exact nightly version, we update
+        // both at the same time.
+
+        let status = process::Command::new("rustup")
```
It definitely feels wrong to update the nightly in behalf of the user.

fitzgen replied:
Unfortunately, this is the only way for the latest `rustfmt` to work :-/

emilio commented on tests/tests.rs:
```diff
@@ -5,42 +5,132 @@ extern crate shlex;
+        assert!(status.success(), "should run `rustup update nightly` OK");
+
+        let status = process::Command::new("rustup")
+            .args(&["run", "nightly", "cargo", "install", "-f", "rustfmt-nightly"])
```
Ditto...

emilio commented:
(Of course this is just for contributors, but still)

emilio commented:
Oh, also @glyn reported when building on top of this branch that he got unexpected test expectation changes, so we should probably get that sorted out.

fitzgen commented:
@glyn, do you have the latest `rustfmt` installed? Can you give more info about the expectation changes you were seeing?

bors-servo commented:
:umbrella: The latest upstream changes (presumably #953) made this pull request unmergeable. Please resolve the merge conflicts.

glyn commented:
@fitzgen Yes, I believe I was using the latest `rustfmt`. Details in [this gist](https://gist.github.com/glyn/4b0d415fbae36dba3bb5961c188829d4).

fitzgen commented:
> @fitzgen Yes, I believe I was using the latest rustfmt. Details in this gist.

Aha! You need to use `rustfmt-nightly`, which is where `rustfmt` development has moved to. The plain `rustfmt` crate is unmaintained now.

fitzgen commented:
@bors-servo r=emilio

bors-servo commented:
:pushpin: Commit 46f74c1 has been approved by `emilio`

<!-- @bors-servo r=emilio 46f74c189c95fc6a41daf8ded741e0cf5f5bed92 -->

bors-servo commented:
:hourglass: Testing commit 46f74c189c95fc6a41daf8ded741e0cf5f5bed92 with merge b13f0766e28cf75dd04f358bf64d87ea96b111c3...

bors-servo commented:
:sunny: Test successful - [status-travis](https://travis-ci.org/rust-lang-nursery/rust-bindgen/builds/273025720?utm_source=github_status&utm_medium=notification)
Approved by: emilio
Pushing b13f0766e28cf75dd04f358bf64d87ea96b111c3 to master...

glyn commented:
@fitzgen `rm ~/.cargo/bin/rustfmt` fixed the problem, thanks.
